File: C++/wildEngimon.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <iterator> 
#include <set>
#include <stdio.h>      
#include <stdlib.h>
#include <time.h> 
#include "skill.hpp"
#include "engimon.hpp"
#include "wildEngimon.hpp"

#include <vector>

using namespace std;

engimon wildEngimon::wildNullEngimon;

wildEngimon::wildEngimon(string sp, int lvl) : engimon (sp, &wildNullEngimon, &wildNullEngimon, sp, lvl, 0)
{
	/* Random Module*/
	srand(time(NULL));
	int randomNumber = rand() %  skill::skill_database.size() + 1;

	// /* Inisialisasi*/
	std::vector<string>::iterator p1;
	std::vector<string>::iterator p2;

	/*Cari 1 skill yang compatible*/
	auto itr = skill::skill_database.begin();
	while (randomNumber > 0)
	{
		if (itr == skill::skill_database.end())
		{
			itr = skill::skill_database.begin();
		}

		/*Fail*/
		vector<string> elementSkill = (*itr).getElements();

		//Find skill dengan elemen yang sesuai
		//p1 = std::find(elementSkill.begin(), elementSkill.end(), this->element1);
		//p2 = std::find(elementSkill.begin(), elementSkill.end(), this->element2);
		p1 = elementSkill.end();
		p2 = elementSkill.end();

		for (auto i = elementSkill.begin(); i != elementSkill.end(); ++i)
		{
			if (*i == this->element1)
			{
				p1 = i;
				break;
			}

			if (*i == this->element2)
			{
				p2 = i;
				break;
			}
		}



		// cout << *elementSkill.begin();
		
		if (p1 != elementSkill.end() || p2 != elementSkill.end())
		{
			randomNumber--;
			if (randomNumber <= 0)
			{
				break;
			}
		}

		if (randomNumber > 0)
		{
			++itr;
		}

		
	}
	try
	{
		this->learnMove((*itr).getSkillName());
	}
	catch (int e)
	{

	}
}

wildEngimon::wildEngimon(const wildEngimon& w) : engimon(w){}

wildEngimon::~wildEngimon(){}

// engimon& wildEngimon::operator=(const wildEngimon& w)
// {
// 	engimon e = new engimon(w.name, w.parent1, w.parent2, w.species, w.level, w.experience);
// 	for (int i = 0; i < 4; ++i)
// 	{
// 		e.setSkill(w.moves[i],i,1);
// 	}
// 	return e;
// }

engimon& wildEngimon::operator=(const wildEngimon& w)
{
	delete[] moves;
	this->name = w.name;
	this->parent1 = w.parent1;
	this->parent2 = w.parent2;
	this->species = w.species;
	this->moves = new skill[4];
	for (int i = 0; i < 4; ++i)
	{
		this->moves[i] = w.moves[i];
	}
	this->element1 = w.element1;
	this->element2 = w.element2;
	this->level = w.level;
	this->experience = w.experience;
	this->cummulative_experience = w.cummulative_experience;
	this->active = w.active;
	return *this;
}

void wildEngimon::learnMove(string move)
{
	auto itr = skill::skill_database.begin();
	for (itr = skill::skill_database.begin(); itr != skill::skill_database.end(); ++itr)
	{
		if ((*itr).getSkillName() == move)	
		{
			break;
		}
	}
    
    if (itr == skill::skill_database.end())
    {
    	throw -1; //"Error, skill tidak ditemukan"
    }

    if ((*itr).getEngimonSpecies() != this->species && (*itr).getEngimonSpecies() != "0")
    {
    	throw -2; //"Error, engimon tidak sesuai"
    }

    vector<string> move_elements = (*itr).getElements();

    auto element1_checker = find(move_elements.begin(), move_elements.end(), this->element1);
    auto element2_checker = find(move_elements.begin(), move_elements.end(), this->element2);

    if (element1_checker == move_elements.end() && element2_checker == move_elements.end())
    {
    	throw -3; //"Error, skill tidak compatible dengan elemen engimon"
    }

    int i = 0;
    for (i = 0; i < 4; ++i)
    {
    	if (this->moves[i].getSkillName() == move)
    	{
    		throw -4; //"Error, skill sudah ada"
    	}
    	if (this->moves[i].getSkillName() == "null")
    	{
    		break;
    	}
    }

    while (!(0 <= i && i <= 3))
    {
    	this->printMoves();
    	cout << "Masukkan move yang ingin direplace" << endl;
    	cin >> i;
    }

    this->moves[i] = (*itr);
}

File: Java/wild_engimon.java
import java.util.Random;

public class wild_engimon extends engimon{

    public static int playerHighestPokemon = 30;
    private static species randSpecies = database_engimon.random();
    private static int randLevel = randomInteger(playerHighestPokemon,playerHighestPokemon+10) ;
    private static int extraMove = randomInteger(0,1);


    public static void setPlayerHighestPokemon(int playerHighestPokemon) {
        wild_engimon.playerHighestPokemon = playerHighestPokemon;
    }

    private static void randomize()
    {
        randSpecies = database_engimon.random();
        if (playerHighestPokemon + 10 < 100)
        {
            randLevel = randomInteger(playerHighestPokemon, playerHighestPokemon+10);
        }
        else
        {
            randLevel = randomInteger(playerHighestPokemon, 100);
        }
    }

    private static void randomize(String tiletype)
    {
        try
        {
            randSpecies = database_engimon.random(tiletype);
            if (playerHighestPokemon + 10 < 100)
            {
                randLevel = randomInteger(playerHighestPokemon, playerHighestPokemon+10);
            }
            else
            {
                randLevel = randomInteger(playerHighestPokemon, 100);
            }
        }
        catch (NullPointerException n)
        {
            n.printStackTrace();
        }
        catch (Exception e)
        {
            System.out.println(e.getMessage());
        }
    }

    //Return a random int between start <= x <= end
    private static int randomInteger(int start, int end)
    {
        Random rand = new Random();

        int rand_int = rand.nextInt(end-start+1);
        return rand_int + start;
    }

    public wild_engimon() throws Exception
    {
        //Bisa terjadi null
        wild_engimon.randomize();
        this.Initialize(randSpecies.speciesName, null, null, randSpecies.speciesName, randLevel,0);
        assert (this != null);
        assert (this.moves != null);
        if (wild_engimon.extraMove == 1)
        {
            this.learnMove(database_skill.random(randSpecies.element1,randSpecies.element2).getSkillName());
        }
        this.life = 1;
    }

    public wild_engimon(String species, int level, int extraMove) throws  Exception
    {
        this.Initialize(species, null, null, species, level,0);
        species s = database_engimon.find(species);
        assert (this != null);
        assert (this.moves != null);
        if (extraMove == 1)
        {
            String skillName = database_skill.random(s.element1,s.element2).getSkillName();
            this.learnMove(skillName);
        }
        this.life = 1;
    }


    public wild_engimon(String tiletype) throws Exception
    {
        wild_engimon.randomize(tiletype);
        this.Initialize(randSpecies.speciesName, null, null, randSpecies.speciesName, randLevel,0);
        assert (this != null);
        assert (this.moves != null);
        if (wild_engimon.extraMove == 1)
        {
            this.learnMove(database_skill.random(randSpecies.element1,randSpecies.element2).getSkillName());
        }
        this.life = 1;
    }

    public void gainExp(int exp)
    {
        this.experience = exp % 100;
        this.level = this.level + exp/100;
        this.cummulative_experience = this.cummulative_experience + exp;
    }
    public void setLevel(int lv)
    {
        this.level = lv;
        this.cummulative_experience = this.level + this.experience;
    }

    public void learnMove(String move)
    {
        if (database_skill.isValid(move,this))
        {
            int idx = -1;
            for (int i = 0; i < 4; i++) {

                if (moves[i] == null)
                {
                    idx = i;
                    break;
                }

                if (moves[i].getSkillName().equals(move))
                {
                    break;
                }


            }

            if (idx == 3)
            {

            }
            else if(idx == -1)
            {

            }
            else
            {
                moves[idx] = new skill(database_skill.find(move));
            }
        }
    }
}
